client/state/utils.js

```diff
@@ -382,6 +301,73 @@ export const withoutPersistence = reducer => {
 	return wrappedReducer;
 };
 
+/**
+ * Returns a reducer function with state calculation determined by the result
+ * of invoking the handler key corresponding with the dispatched action type,
+ * passing both the current state and action object. Defines default
+ * serialization (persistence) handlers based on the presence of a schema.
+ *
+ * @param  {*}        initialState   Initial state
+ * @param  {Object}   customHandlers Object mapping action types to state action handlers
+ * @param  {?Object}  schema         JSON schema object for deserialization validation
+ * @return {Function}                Reducer function
+ */
+export function createReducer( initialState, customHandlers, schema ) {
+	let handlers = customHandlers;
+
+	// When custom serialization behavior is provided, we assume that it may
+	// involve heavy logic (mapping, converting from Immutable instance), so
+	// we cache the result and only regenerate when state has changed.
+	const customSerializeHandler = customHandlers[ SERIALIZE ];
+	if ( customSerializeHandler ) {
+		let lastState, lastSerialized;
+		const cachedSerializeHandler = ( state, action ) => {
+			if ( state === lastState ) {
+				return lastSerialized;
+			}
+
+			const serialized = customSerializeHandler( state, action );
+			lastState = state;
+			lastSerialized = serialized;
+			return serialized;
+		};
+		handlers = {
+			...handlers,
+			[ SERIALIZE ]: cachedSerializeHandler,
+		};
```

Comment:
Yes, it's only to avoid mutating the `customHandlers` object passed as argument. I'll add an explaining comment.

I'm not sure if memoizing the SERIALIZE handler still has much value today -- we almost completely abandoned ImmutableJS and most of other serialization is very fast, e.g., extracting `data` and `options` properties from `QueryManager` objects etc.

Reply:
It doesn't seem unreasonable to leave that up to the `SERIALIZE` handler's implementation.

Reply:
There are only ~10 files that directly make use of SERIALIZE. I don't think we should have special case support for them -- I agree with you two that they could implement their own memoization if necessary

Reply:
And these are the only 7 *.js files with both `createReducer` _and_ `SERIALIZE` to look at:

```
client/state/posts/reducer.js
-> client/state/utils.js <- no action required
client/state/terms/reducer.js
client/state/test/utils.js
client/state/themes/reducer.js
client/state/reader/feeds/reducer.js
client/state/reader/sites/reducer.js
client/state/reader/follows/reducer.js
```

An audit should be quick, and we can drop the special case.

Reply:
> An audit should be quick, and we can drop the special case.

💯. Totally agree.  We can merge this first as its already a good improvement

Reply:
@samouri @sirreal I pushed a commit where I removed the SERIALIZE caching, together with the respective test. `createReducer` got really simple now :tada: